Make image compression format configurable

gabriel-samfira opened the pull request:
# Make image compression format configurable

This change adds a new flag called --image_compression_format which allows us to output the final VM image as one of the supported formats: bz2 (default), gz, zip or none

If the compression format is "none" or "", the image will not be compressed.

This change also enables the CI to output OpenStack images using ```gzip``` compression instead of ```bz2```. 

Fixes https://github.com/flatcar-linux/Flatcar/issues/575

## How to use

Simply add ```--image_compression_format=gz``` when calling ```image_to_vm.sh```. Leaving this option unset will preserve the current behavior.

## Testing done

These changes were tested by specifying all supported formats as well as arbitrary bogus strings. The results were as expected. Success on any of the supported formats, error on anything else. Image archive as well as legacy digest files were properly generated.

- [x] Changelog entries added in the respective `changelog/` directory (user-facing change, bug fix, security fix, update)

pothos commented:
Thanks, looks good on a first glance but the images are also used by the `flatcar-install` script and it won't find them anymore with the new extension. Same for users who have download scripts which will break. I suggest to rework this to support both formats at the same time, i.e., the flag should accept a list of compression formats and then we pass `.gz,.bz2` for OpenStack images to generate the gzip images additionally.

jepio commented on build_library/release_util.sh:
```diff
@@ -42,6 +40,38 @@ DEFINE_string sign "" \
+    "zip")
+        IMAGE_ZIPPER="pigz --keep --zip"
+        ;;
+    *)
```
could you also add support for zstd?

pothos replied:
Had the same thought but we can also add it later and not as part of this PR

jepio replied:
make sense

gabriel-samfira replied:
Seems like the binary is already in the sdk container, so why not? :D

gabriel-samfira commented:
> Thanks, looks good on a first glance but users may have download scripts which will break. I suggest to rework this to support both formats at the same time, i.e., the flag should accept a list of compression formats and then we pass `.gz,.bz2` for OpenStack images to generate the gzip images additionally.

Looking at where to best add this. Will probably update the PR tomorrow morning.

> Edit: `flatcar-install` does not consume the `.img` files, so no change needed there, but the https://github.com/flatcar-linux/flatcar-docs have references to `openstack_image.img.bz2`

Will also update the docs.

gabriel-samfira commented:
@pothos Updated the PR. The change adds the ability to specify multiple compression formats as a comma separated value. The compression logic was moved outside of the upload function.

I also added ```zstd``` as an option. Please take another look.

pothos commented on build_library/release_util.sh:
```diff
@@ -42,6 +40,43 @@ DEFINE_string sign "" \
+        IMAGE_ZIPPER="pigz --keep --zip"
+        ;;
+    "zstd")
+       IMAGE_ZIPPER="zstd --format=zstd -k -q -f --no-progress -o ${filepath}.${compression_format}"
```
```suggestion
       IMAGE_ZIPPER="zstd --format=zstd -k -q --no-progress -o ${filepath}.${compression_format}"
```
I see that `-f` is specified below. (Any of the two places is fine but it should be consitent.)

gabriel-samfira replied:
Oops. Missed that one. Fixed.

pothos commented:
> Sounds good. Should I do that, or is there a CI that does that? Apologies, I am a newbie :D.

We have to manually start the CI for now, will do soon :D

pothos commented:
Something is wrong because `flatcar_production_image.bin.bz2.sig` wasn't generated

pothos commented:
Ah, `flatcar_production_image.bin.sig` was generated instead

gabriel-samfira commented:
> Something is wrong because `flatcar_production_image.bin.bz2.sig` wasn't generated and the `.bz2` is missing, too

Are there any CI logs I can check? The link doesn't seem to be accessible.

> Ah, `flatcar_production_image.bin.sig` and `flatcar_production_image.bin` got uploaded instead

Ack. Looking into it now.

pothos commented:
The `build_image` step is responsible for the plain image: https://github.com/flatcar-linux/scripts/blob/main/jenkins/images.sh#L119
We have to test both, not only the `image_to_vm.sh` case

pothos commented:
The changed `upload_image` function is called at various places, which is probably where the failures come from.

gabriel-samfira commented:
> The `build_image` step is responsible for the plain image: https://github.com/flatcar-linux/scripts/blob/main/jenkins/images.sh#L119
> We have to test both, not only the `image_to_vm.sh` case

yup. Saw it. Testing now, but iteration time is slow :D. Will ping in the comments once done.

gabriel-samfira commented:
Made some changes that also takes care of compressing updates and should handle cases where ```upload_image()``` is called. The change makes the code a bit more verbose and repetitive, but it decouples compressing images from uploading them, giving the caller clearer expectations of what to expect when calling ```upload_image()```.

Currently, from what I see, ```upload_image()``` is used to upload not only images, but also other types of files. It also is responsible for compressing images. The need to be able to compress using multiple formats and retain ```DIGESTS``` files in a way that legacy scripts won't fail, has made this a bit more verbose than I expected.

If you prefer that I rework this a bit, let me know. Any suggestions are of course welcome.

gabriel-samfira commented:
> Can you have a look at the suggestions here (mainly about quoting and the changelog entry)?

Yes. Of course! First thing in the morning.

pothos commented on build_library/build_image_util.sh:
```diff
@@ -58,10 +58,28 @@ extract_update() {
+  compress_disk_images image_files compressed_images
+
+  # Upload compressed image
+  upload_image -d "${digest_path}" ${compressed_images[@]}
```
What if the image wasn't compressed? This should upload the `image_files` entry then. I suggest you add an "else" case to `compress_disk_images` where the check for the img/vmdk etc extension is done and add the plain file otherwise. Now that means the naming of "compressed_images" is a bit wrong and we can rather call it "images_to_upload".

pothos replied:
Ok, then I suggest to have two outputs, one is compressed, one is the list of skipped files and then the `upload_image` command can pass the concatenation of both and this also helps to remove the duplication in `vm_image_util.sh`, or?

gabriel-samfira replied:
let's give that a shot.

gabriel-samfira replied:
updated

pothos commented on build_library/modify_image_util.sh:
```diff
@@ -83,7 +83,13 @@ finish_modify_image() {
+    image_files+=( "${DST_IMAGE}" )
+    compress_disk_images image_files compressed_images
+
+    upload_image "${compressed_images[@]}"
```
Same here, the image may not get compressed if the extension mismatches

gabriel-samfira replied:
Is there any situation in which this would not be a compress-able image (with a matching extension)?

pothos commented on build_library/prod_image_util.sh:
```diff
@@ -138,17 +138,34 @@ EOF
+
+  local image_files=( "${BUILD_DIR}/${image_name}" )
+  declare -a compressed_images
+  compress_disk_images image_files compressed_images
```
Same here

pothos commented on build_library/vm_image_util.sh:
```diff
@@ -1199,45 +1199,54 @@ vm_cleanup() {
+    declare -a digest_uploads
+
+    for file in "${VM_GENERATED_FILES[@]}";do
+        if [[ "${file}" =~ \.(img|bin|vdi|vhd|vmdk)$ ]]; then
```
This duplicates the logic in `compress_disk_images` and I think it's better to move it there completely as suggested above

pothos replied:
Ok, this legacy logic is a bit complicated, thanks for your works there. I think it's good now, or?

gabriel-samfira replied:
Yup. Should be ok. The changes to ```compress_disk_image()``` makes is also populate an "extra_files" array with whatever doesn't match that image pattern. Which callers can then concatenate and upload if they wish.

gabriel-samfira commented:
Added quotes to variables, updated changelog.

I combed through the comments here, but I do not see your suggestions regarding quotes and changelog. Perhaps they were not published?

pothos commented:
> Added quotes to variables, updated changelog.
> 
> I combed through the comments here, but I do not see your suggestions regarding quotes and changelog. Perhaps they were not published?

Oh, sorry, that was it

pothos commented:
The last completed successfully but I didn't have a close look at the list of resulting files.
I've started a new test run now with these changes and will paste the file lists.

gabriel-samfira commented:
> The last completed successfully but I didn't have a close look at the list of resulting files.

should be the same files, plus an extra ```.DIGESTS``` file with the name of the uncompressed file when building an image. Here is the output on my machine:

For:

```bash
./image_to_vm.sh --upload_root rsync://ubuntu@localhost:/mnt/test_upload --upload
```
The result in both cases (main branch and this one):

```bash
ubuntu@flatcar:~/scripts$ ls -l /mnt/test_upload/boards/amd64-usr/3139.0.0+2022-03-03-1016.compress_to_image
total 381328
-rw-r--r-- 1 ubuntu ubuntu      1017 Mar  3 10:44 flatcar_production_qemu.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu       535 Mar  3 10:44 flatcar_production_qemu.README
-rwxr-xr-x 1 ubuntu ubuntu      7518 Mar  3 10:44 flatcar_production_qemu.sh
-rw-r--r-- 1 ubuntu ubuntu 390453496 Mar  3 10:44 flatcar_production_qemu_image.img.bz2
-rw-r--r-- 1 ubuntu ubuntu      1017 Mar  3 10:44 flatcar_production_qemu_image.img.bz2.DIGESTS
ubuntu@flatcar:~/scripts$ ls -l /mnt/test_upload/boards/amd64-usr/3139.0.0+2022-03-03-0918.to_image/
total 381792
-rw-r--r-- 1 ubuntu ubuntu      1017 Mar  3 10:13 flatcar_production_qemu.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu       535 Mar  3 10:13 flatcar_production_qemu.README
-rwxr-xr-x 1 ubuntu ubuntu      7518 Mar  3 10:13 flatcar_production_qemu.sh
-rw-r--r-- 1 ubuntu ubuntu 390928218 Mar  3 10:13 flatcar_production_qemu_image.img.bz2
-rw-r--r-- 1 ubuntu ubuntu      1017 Mar  3 10:13 flatcar_production_qemu_image.img.bz2.DIGESTS
```

For:

```bash
./build_image --upload_root rsync://ubuntu@localhost:/mnt/test_upload --upload
```

The result for both branches:

```bash
ubuntu@flatcar:~/scripts$ ls -l /mnt/test_upload/boards/amd64-usr/3139.0.0+2022-03-03-0918.bk/
total 1063860
-rw-r--r-- 1 ubuntu ubuntu 391670006 Mar  3 09:39 flatcar_production_image.bin.bz2
-rw-r--r-- 1 ubuntu ubuntu      3159 Mar  3 09:39 flatcar_production_image.bin.bz2.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu    717304 Mar  3 09:39 flatcar_production_image.grub
-rw-r--r-- 1 ubuntu ubuntu   1285928 Mar  3 09:39 flatcar_production_image.shim
-rw-r--r-- 1 ubuntu ubuntu  52289048 Mar  3 09:39 flatcar_production_image.vmlinuz
-rw-r--r-- 1 ubuntu ubuntu   2306240 Mar  3 09:39 flatcar_production_image_contents.txt
-rw-r--r-- 1 ubuntu ubuntu    152830 Mar  3 09:39 flatcar_production_image_kernel_config.txt
-rw-r--r-- 1 ubuntu ubuntu   1843726 Mar  3 09:39 flatcar_production_image_licenses.json
-rw-r--r-- 1 ubuntu ubuntu     10366 Mar  3 09:37 flatcar_production_image_packages.txt
-rw-r--r-- 1 ubuntu ubuntu     20321 Mar  3 09:39 flatcar_production_image_pcr_policy.zip
-rw-r--r-- 1 ubuntu ubuntu 288111708 Mar  3 09:39 flatcar_production_update.bin.bz2
-rw-r--r-- 1 ubuntu ubuntu       345 Mar  3 09:39 flatcar_production_update.bin.bz2.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu  11547194 Mar  3 09:40 flatcar_production_update.zip
-rw-r--r-- 1 ubuntu ubuntu       333 Mar  3 09:40 flatcar_production_update.zip.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu 339375926 Mar  3 09:40 flatcar_test_update.gz
-rw-r--r-- 1 ubuntu ubuntu       312 Mar  3 09:40 flatcar_test_update.gz.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu       185 Mar  3 09:40 version.txt
-rw-r--r-- 1 ubuntu ubuntu       279 Mar  3 09:40 version.txt.DIGESTS
ubuntu@flatcar:~/scripts$ ls -l /mnt/test_upload/boards/amd64-usr/3139.0.0+2022-03-03-1016.compress_image/
total 1061832
-rw-r--r-- 1 ubuntu ubuntu      3159 Mar  3 10:37 flatcar_production_image.bin.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu 390943703 Mar  3 10:36 flatcar_production_image.bin.bz2
-rw-r--r-- 1 ubuntu ubuntu      3159 Mar  3 10:37 flatcar_production_image.bin.bz2.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu    717304 Mar  3 10:36 flatcar_production_image.grub
-rw-r--r-- 1 ubuntu ubuntu   1285928 Mar  3 10:36 flatcar_production_image.shim
-rw-r--r-- 1 ubuntu ubuntu  52289048 Mar  3 10:36 flatcar_production_image.vmlinuz
-rw-r--r-- 1 ubuntu ubuntu   2306240 Mar  3 10:36 flatcar_production_image_contents.txt
-rw-r--r-- 1 ubuntu ubuntu    152830 Mar  3 10:36 flatcar_production_image_kernel_config.txt
-rw-r--r-- 1 ubuntu ubuntu   1843726 Mar  3 10:36 flatcar_production_image_licenses.json
-rw-r--r-- 1 ubuntu ubuntu     10366 Mar  3 10:35 flatcar_production_image_packages.txt
-rw-r--r-- 1 ubuntu ubuntu     20322 Mar  3 10:36 flatcar_production_image_pcr_policy.zip
-rw-r--r-- 1 ubuntu ubuntu       345 Mar  3 10:37 flatcar_production_update.bin.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu 287332590 Mar  3 10:37 flatcar_production_update.bin.bz2
-rw-r--r-- 1 ubuntu ubuntu       345 Mar  3 10:37 flatcar_production_update.bin.bz2.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu  11547194 Mar  3 10:38 flatcar_production_update.zip
-rw-r--r-- 1 ubuntu ubuntu       333 Mar  3 10:38 flatcar_production_update.zip.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu 338802477 Mar  3 10:38 flatcar_test_update.gz
-rw-r--r-- 1 ubuntu ubuntu       312 Mar  3 10:38 flatcar_test_update.gz.DIGESTS
-rw-r--r-- 1 ubuntu ubuntu       185 Mar  3 10:38 version.txt
-rw-r--r-- 1 ubuntu ubuntu       279 Mar  3 10:38 version.txt.DIGESTS
```

If the CI could run all the code paths you guys usually run to properly validate that I have not missed something, that would be great.

pothos commented:
I got this error when building:
```
INFO    build_image: Cleaning up mounts
INFO    build_image: Compressing images
INFO    build_image: Compressing flatcar_developer_container.bin to bz2
cp: cannot stat '/home/sdk/build/images/amd64-usr/developer-9999.9.9+kai-test-pr-233-a1/flatcar_developer_container.bin.DIGESTS': No such file or directory
ERROR   build_image: script called: build_image '--board=amd64-usr' '--output_root=/home/sdk/build/images' '--torcx_root=/home/sdk/build/torcx' 'prodtar' 'container'
ERROR   build_image: Backtrace:  (most recent call is last)
ERROR   build_image:   file build_image, line 173, called: create_dev_container 'flatcar_developer_container.bin' 'container' 'https://mirror.release.flatcar-linux.net/' 'developer' 'coreos-base/coreos-dev'
ERROR   build_image:   file dev_container_util.sh, line 128, called: die_err_trap 'cp "${BUILD_DIR}/${image_name}.DIGESTS" "${digest_file}"' '1'
ERROR   build_image: 
ERROR   build_image: Command failed:
ERROR   build_image:   Command 'cp "${BUILD_DIR}/${image_name}.DIGESTS" "${digest_file}"' exited with nonzero code: 1
flatcar-images-amd64-9999.9.9-kai-test-pr-233
```

gabriel-samfira commented:
> I got this error when building:

Checking. Will update today. Thanks for the debug output!

gabriel-samfira commented:
@pothos any chance there's a guide on how to set up a 3rd party CI that can mirror yours? I'd like to set one up to not have to bug you.

pothos commented:
At the moment we have two setups, one is the inherited using cork, and the new one uses the run_sdk_container script for Docker. The failure happened with the new setup - while not complete I'll attach the Jenkins files here: 
[jenkins.tar.zip](https://github.com/flatcar-linux/scripts/files/8196617/jenkins.tar.zip)

gabriel-samfira commented:
> At the moment we have two setups, one is the inherited using cork, and the new one uses the run_sdk_container script for Docker. The failure happened with the new setup - while not complete I'll attach the Jenkins files here:

Thanks! Currently adding a function for the legacy digests, testing locally and pushing. I managed to run ```image_build amd64```, which worked until it had to upload files to ```bincache@bincache.flatcar-linux.net``` :grin:

gabriel-samfira commented:
Updated the PR. If you could share the error from the previous failure, that would be great.

pothos commented:
I think it worked now! There was a just build failure on arm64 which may be due to needing a rebase.

gabriel-samfira commented:
> I think it worked now! There was a just build failure on arm64 which may be due to needing a rebase.

I rebased on top of main, and squashed most of the commits. Give it another go. Running it locally as well.

pothos reviewed (approved):
Started another test but I think all is covered now, thanks

pothos commented:
(While at it we can also add a gz image for DigitalOcean to address the temporary failures the note in https://www.flatcar.org/docs/latest/installing/cloud/digitalocean/ talks about)

pothos commented on jenkins/vms.sh:
```diff
@@ -113,6 +113,12 @@ for FORMAT in ${FORMATS}; do
 
+  COMPRESSION_FORMAT="bz2"
+
+  if [[ "${FORMAT}" =~ ^(openstack|openstack_mini)$ ]];then
```
```suggestion
  if [[ "${FORMAT}" =~ ^(openstack|openstack_mini|digitalocean)$ ]];then
```

pothos commented on changelog/changes/2022-02-22-configurable-image-compression.md:
```diff
@@ -0,0 +1,2 @@
+- OpenStack: In addition to the `bz2` image, a `gz` compressed image is published. This allows Glance to directly consume the images by simply passing in the URL of the image.
+- SDK: The image compression format is now configurable. Supported formats are: `bz2`, `gz`, `zip`, `none`, `zst`. Selecting the image format can now be done by passing the `--image_compression_formats` option. This flag gets a comma separated list of formats.
```
```suggestion
- DigitalOcean: In addition to the `bz2` image, a `gz` compressed image is published. This helps against hitting the compression timeout that sometimes lets the image import fail.
- SDK: The image compression format is now configurable. Supported formats are: `bz2`, `gz`, `zip`, `none`, `zst`. Selecting the image format can now be done by passing the `--image_compression_formats` option. This flag gets a comma separated list of formats.
```

pothos commented:
Thanks again :)

gabriel-samfira commented:
> Thanks again :)

My pleasure! Thank you for the guidance and patience! :D

pothos commented:
It has been working well so far in Beta and Alpha releases. I'll pick it for the Stable/LTS branches to reduce the build system differences across the branches.
